### PR TITLE
[sea-orm-linter] statement builder with plugin

### PR DESCRIPTION
**Pull request by billy1624**
## PR Info

Drafting a statement builder plugin API. Specifically designed for SeaORM linter :)

CC @kirawi

### CODE REVIEW

File: src/linter.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    fn run(&self, stmt: &AnyStatement) {
+        match stmt {
+            AnyStatement::Insert(stmt) => InsertQueryLinter::run(stmt),
+            AnyStatement::Select(stmt) => SelectQueryLinter::run(stmt),
+            AnyStatement::Update(stmt) => UpdateQueryLinter::run(stmt),
+            AnyStatement::Delete(stmt) => DeleteQueryLinter::run(stmt),
+            _ => {}
+        }
+    }
+}
```
Comment (billy1624) [inline src/linter.rs:33]: Linter should be defined in a separate crate. But for simplicity and demo purpose. I just place it inside SeaORM.

File: src/database/statement.rs [hunk 2]
```diff
@@ -16,9 +16,50 @@ pub struct Statement {
...
+        I: IntoIterator<Item = T>,
+        T: StatementBuilderPlugin,
+    {
+        let stmt = self.build(db_backend);
+        let any_stmt = self.into_any_statement();
+        for plugin in plugins {
+            plugin.run(&any_stmt);
+        }
+        stmt
+    }
```
Comment (billy1624) [inline src/database/statement.rs:34]: Build a SeaQuery statement with linters (plugins)

File: src/database/statement.rs [hunk 3]
```diff
@@ -16,9 +16,50 @@ pub struct Statement {
...
+}
+
+#[derive(Debug)]
+pub enum AnyStatement<'a> {
+    Insert(&'a sea_query::InsertStatement),
+    Select(&'a sea_query::SelectStatement),
+    Update(&'a sea_query::UpdateStatement),
+    Delete(&'a sea_query::DeleteStatement),
```
Comment (billy1624) [inline src/database/statement.rs:42]: Any statement without `dyn`

**Comment by kirawi** [comment]
Thanks! This seems like a better solution than what I was attempting.

File: src/executor/select.rs [hunk 4]
```diff
@@ -423,7 +423,8 @@ where
         C: ConnectionTrait,
     {
         let builder = db.get_database_backend();
-        let stmt = builder.build(&self.query);
+        let linter = QueryLinter::new(db);
+        let stmt = builder.build_with_plugins(&self.query, [linter]);
```
[the remark this answers is not in the document]
Comment (kirawi) [reply src/executor/select.rs:427]: I can think of a couple different approaches, but it's unclear to me what the requirements for a minimal integration point for the linter are. This PR seems to favor an explicit use of the linter by the end user over an implicit approach like my PR did. Is that what we're aiming for?

**Comment by kirawi** [comment]
One issue is that there currently isn't a way to access the sqlx pool through `ConnectionTrait`. Maybe a method could be added to return one?

**Comment by billy1624** [comment]
> One issue is that there currently isn't a way to access the sqlx pool through `ConnectionTrait`. Maybe a method could be added to return one?

Hey @kirawi, related issues:
- https://github.com/SeaQL/sea-orm/issues/753

No active PR at the moment.

But why would you need the underlying SQLx driver? Why not just execute the statement via SeaORM?
- https://www.sea-ql.org/SeaORM/docs/basic-crud/raw-sql#use-raw-query--execute-interface

**Comment by kirawi** [comment]
`sea-schema` seems to require an sqlx pool for schema discovery, e.g. https://docs.rs/sea-schema/latest/sea_schema/sqlite/discovery/struct.SchemaDiscovery.html#method.new

**Comment by tyt2y3** [comment]
I think this is a great starting point, especially the `AnyStatement`
Obviously, we should not `Linter::new()` everywhere. Under SeaORM's lint mode, we should
1) run schema discovery and store it inside the connection pool (`Arc` as it should be immutable)
2) initialize a Linter and again `Arc` it (I expect it to be immutable afterwards)
3) run entity schema check first (they may not be in sync!)
4) during query exec, invoke the linter to do some checking

**Comment by tyt2y3** [comment]
This is the *bare minimum* MVP we are looking for

**Comment by tyt2y3** [comment]
And, there should exist a `Linter` trait to act as an adapter, and that we can expose a macro to magically glue things together (like sea_query driver), so in the end, neither `sea-orm` nor `sea-orm-linter` depends on one another.

**Comment by kirawi** [comment]
Is there a reason `sea-schema` requires an sqlx pool to detect the schema, rather than just a connection?

**Comment by tyt2y3** [comment]
> Is there a reason `sea-schema` requires an sqlx pool to detect the schema, rather than just a connection?

No. We can probably change the API of sea-schema to accept a connection

**Comment by kirawi** [comment]
No worries. I read the docs for `Pool` and a clone is cheap.

**Pull request closed without merging**